oeps/architectural-decisions/oep-XXXX-modular-micro-frontend-domains.rst

```diff
@@ -0,0 +1,209 @@
+
+As this OEP is developed, we will further investigate:
+
+#. Integration strategy
+#. Domain SPAs and their interfaces
+#. Shared dependencies
+#. MFEs as components
+
+Discovery
+=========
+
+* Are there any further top-level architectural elements to consider?
+* What is the relationship of this project to the Modular Learning effort?
+
+  * Both are long-term projects, but beyond that, is there anything else that could be coordinated between them?
+  * Should this project target the modular codebase exclusively, ignoring ``edx-platform``?  (Relevant if the APIs are incompatible between them.)
+
+Integration strategy
+====================
+
+To be defined.
+
+Discovery
+---------
+
+At time of writing these appear to commonly recommended strategies for integration of MFEs into modular container apps:
+
+* `Run-time integration via JavaScript <https://martinfowler.com/articles/micro-frontends.html#Run-timeIntegrationViaJavascript>`_
+* `Run-time integration via Web Components <https://martinfowler.com/articles/micro-frontends.html#Run-timeIntegrationViaWebComponents>`_
+* `Webpack Module Federation <https://webpack.js.org/concepts/module-federation/>`_
+* `single-spa <https://single-spa.js.org/>`_
+* `Piral <https://github.com/smapiot/piral>`_
+
+With these in mind:
+
+* What are the pros and cons of each integration option, in particular in terms of:
+
+  * Cost of conversion: how easy is it to convert existing MFEs to the new paradigm?
+  * Cost of integration: compare the level of MFE independence.
+  * Hosting of build artifacts: is it possible to host build artifacts separately on object storage or CDN?
+  * Lazy loading: is it possible to load MFEs and other components lazily?
+  * Efficiency of development: consider the level of boiler-plate code.
+  * User impact: compare the benefits and losses from the point of view of the browser.
+  * Maturity: how reliable is the codebase?  Is it maintained actively?  Can it be relied upon to exist in 5 years?
+  * Project health: is the project active (measure commit activity)?  How active is the community?
+
+* Recommend a single choice between the integration options, or suggest another, better alternative.
+
+Domain SPAs and their interfaces
+================================
+
+To be defined.
+
+Discovery
+---------
+
+* What UX elements should go into the domain's containing app? The header? Footer?
+* Define further container SPA and interface characteristics.
+
+Shared dependencies
+===================
+
+To be defined.
+
+Discovery
+---------
+
+* What dependencies should be shared between all MFEs?  (React, Redux, ...?)
```

Comment:
`@edx/paragon`? I would imagine having a single Paragon version should lead to a more consistent UX in terms of design and functionality across the domain.

Reply:
The Paragon library supports tree shaking, so it only includes the necessary code for the components an MFE is actually using, which should mitigate this concern. As far as I can tell, I don't believe that existing tree shaking behavior would be impacted even with a single copy of Paragon shared across multiple MFEs in a domain app.

Reply:
Aha but what about the case where the MFE is expected to use a shared version of paragon, which I expect suppose to be available at runtime? _Below is a scinario of how I imagine but I might be wrong_ 

- Paragon has compoenets: (A, B, C, D, E, F, G,H) 
- MFE-1: depends on (A, B, C E, F)
- MFE-2 depends on (C, D, G). 

The de-facto situation wihhout having the host app...etc is that webapck suppose to to do tree shakeing when building MFE-1 and MFE-2, such that they will denepd on a version of paragon that has a subset of paragon compoenents. 

However in the host/container app siutation, then we are thinking we have a copy of paragon for all other MFEs, _so we don't load paragon everytime a user change MFE_? If yes then which componenet do we initially include in the host app, my top of head suggestion can be: 

1. Just include all of them, _if the added complexity doesn't worth the saved bandwidth..._ 
1. Include an _Union of all MFEs paragon compoenets_  i.e. $MFE1 \cup MFE2... \cup MFEn = PU$
  - The more MFEs we have the closet $PU$ gets to $P$ where $P$ is a set of all the compoenet 
2. Incldue an _Intersection of all MFEs paragon compoenets_ i.e $MFE1 \cap MFE2 ...\cap MFEn = PI$
  - The more MFEs we have then probably the less compoenents $PI$ would contain. 
  - This path of course would add a bit complexity beacus  loading a particualr  MFE, then we need to load the difference in compoenets, i.e $MFE1 - PI$  _The component that are in $MFE1$ but not in $PI$
    - One other example is load each compoenet just it loads, but that of course cause that for each componenet in the MFE that doesn't exists in the third set, we to have a roundtrip to the server